File: packages/arcis-python/arcis/middleware/sensitive_paths.py
"""v1.7 W2 wire-up. Blocks well-known scanner probe paths.

Three buckets, almost never legitimate on a typical app:
    1. Dotfile / VCS leaks   /.env, /.git/*, /.svn/*, /.aws/, ...
    2. PHP/Wordpress probes  /wp-admin, /wp-login.php, /phpmyadmin, ...
    3. Diagnostic endpoints  /server-status, /phpinfo.php, /info.php, ...

Apps with legitimate overlapping routes (actual WordPress site, custom
/admin panel) opt out via Arcis(scanner_paths=False) / Arcis class +
ArcisMiddleware(scanner_paths=False).
"""

from __future__ import annotations

import re
from typing import List, Optional


SENSITIVE_PATH_PATTERNS: List[re.Pattern] = [
    # Dotfile / VCS leaks — should never be served on any app.
    re.compile(r"^/\.env(\.|/|$)", re.IGNORECASE),
    re.compile(r"^/\.git(/|$)", re.IGNORECASE),
    re.compile(r"^/\.svn(/|$)", re.IGNORECASE),
    re.compile(r"^/\.hg(/|$)", re.IGNORECASE),
    re.compile(r"^/\.bzr(/|$)", re.IGNORECASE),
    re.compile(r"^/\.aws(/|$)", re.IGNORECASE),
    re.compile(r"^/\.ssh(/|$)", re.IGNORECASE),
    re.compile(r"^/\.htaccess$", re.IGNORECASE),
    re.compile(r"^/\.htpasswd$", re.IGNORECASE),
    re.compile(r"^/\.npmrc$", re.IGNORECASE),
    re.compile(r"^/\.dockerenv$", re.IGNORECASE),

    # WordPress + PHP probes.
    re.compile(r"^/wp-admin(/|$)", re.IGNORECASE),
    re.compile(r"^/wp-login\.php$", re.IGNORECASE),
    re.compile(r"^/wp-config\.php$", re.IGNORECASE),
    re.compile(r"^/wordpress/wp-(admin|login)", re.IGNORECASE),
    re.compile(r"^/xmlrpc\.php$", re.IGNORECASE),

    # Generic admin / DB-admin probes.
    re.compile(r"^/admin/?$", re.IGNORECASE),
    re.compile(r"^/administrator/?$", re.IGNORECASE),
    re.compile(r"^/admin\.php$", re.IGNORECASE),
    re.compile(r"^/phpmyadmin(/|$)", re.IGNORECASE),
    re.compile(r"^/pma(/|$)", re.IGNORECASE),
    re.compile(r"^/myadmin(/|$)", re.IGNORECASE),
    re.compile(r"^/dbadmin(/|$)", re.IGNORECASE),
    re.compile(r"^/adminer\.php$", re.IGNORECASE),

    # Diagnostic / info-leak endpoints.
    re.compile(r"^/phpinfo\.php$", re.IGNORECASE),
    re.compile(r"^/info\.php$", re.IGNORECASE),
    re.compile(r"^/test\.php$", re.IGNORECASE),
    re.compile(r"^/shell\.php$", re.IGNORECASE),
    re.compile(r"^/server-status$", re.IGNORECASE),
    re.compile(r"^/server-info$", re.IGNORECASE),

    # Backup / dump leaks.
    re.compile(r"^/backup(\.|/)", re.IGNORECASE),
    re.compile(r"^/dump\.sql$", re.IGNORECASE),
    re.compile(r"^/database\.sql$", re.IGNORECASE),
]


def detect_sensitive_path(
    path: str,
    patterns: Optional[List[re.Pattern]] = None,
) -> Optional[str]:
    """Test a URL path against the sensitive-path list.

    Returns the first matching pattern's source string for logging /
    telemetry, or None if no match.
    """
    pats = patterns if patterns is not None else SENSITIVE_PATH_PATTERNS
    for pat in pats:
        if pat.search(path):
            return pat.pattern
    return None

File: packages/arcis-node/src/middleware/sensitive-paths.ts
/**
 * @module @arcis/node/middleware/sensitive-paths
 * v1.7 W2 wire-up. Blocks well-known scanner probe paths.
 *
 * Sensitive paths fall into three buckets and almost never have a
 * legitimate reason to be served by a typical app:
 *   1. Dotfile / VCS leaks   `/.env`, `/.git/*`, `/.svn/*`, `/.aws/`, ...
 *   2. PHP/Wordpress probes  `/wp-admin`, `/wp-login.php`, `/phpmyadmin`, ...
 *   3. Diagnostic endpoints  `/server-status`, `/phpinfo.php`, `/info.php`, ...
 *
 * Apps with legitimate routes that overlap (an actual WordPress site,
 * a custom `/admin` panel) opt out via `arcis({ scannerPaths: false })`
 * or pass a custom matcher list.
 */

import type { Request, Response, NextFunction, RequestHandler } from 'express';

export interface ScannerPathsOptions {
  /** HTTP status for blocked probes. Default: 403 */
  statusCode?: number;
  /** Error message for blocked probes. Default: 'Access denied.' */
  message?: string;
  /**
   * Custom matchers (overrides the default list entirely). Each entry
   * is matched against `req.path` with `.test()`. To EXTEND the default
   * list use `[...SENSITIVE_PATH_PATTERNS, /your-extra/]`.
   */
  patterns?: RegExp[];
}

/**
 * Default sensitive-path patterns. Each is a case-insensitive prefix
 * match anchored to the start of the request path. Conservative on
 * shapes that have any reasonable legit use; aggressive on dotfile and
 * VCS-leak paths that should never be served.
 */
export const SENSITIVE_PATH_PATTERNS: ReadonlyArray<RegExp> = Object.freeze([
  // Dotfile / VCS leaks — should never be served on any app.
  /^\/\.env(\.|\/|$)/i,
  /^\/\.git(\/|$)/i,
  /^\/\.svn(\/|$)/i,
  /^\/\.hg(\/|$)/i,
  /^\/\.bzr(\/|$)/i,
  /^\/\.aws(\/|$)/i,
  /^\/\.ssh(\/|$)/i,
  /^\/\.htaccess$/i,
  /^\/\.htpasswd$/i,
  /^\/\.npmrc$/i,
  /^\/\.dockerenv$/i,

  // WordPress + PHP probes — legit only on those platforms.
  /^\/wp-admin(\/|$)/i,
  /^\/wp-login\.php$/i,
  /^\/wp-config\.php$/i,
  /^\/wordpress\/wp-(admin|login)/i,
  /^\/xmlrpc\.php$/i,

  // Generic admin / DB-admin probes.
  /^\/admin\/?$/i,
  /^\/administrator\/?$/i,
  /^\/admin\.php$/i,
  /^\/phpmyadmin(\/|$)/i,
  /^\/pma(\/|$)/i,
  /^\/myadmin(\/|$)/i,
  /^\/dbadmin(\/|$)/i,
  /^\/adminer\.php$/i,

  // Diagnostic / info-leak endpoints.
  /^\/phpinfo\.php$/i,
  /^\/info\.php$/i,
  /^\/test\.php$/i,
  /^\/shell\.php$/i,
  /^\/server-status$/i,
  /^\/server-info$/i,

  // Backup / dump leaks.
  /^\/backup(\.|\/)/i,
  /^\/dump\.sql$/i,
  /^\/database\.sql$/i,
]);

/**
 * Test a URL path against the sensitive-path list. Returns the first
 * matching pattern's source for logging/telemetry, or `null`.
 */
export function detectSensitivePath(
  path: string,
  patterns: ReadonlyArray<RegExp> = SENSITIVE_PATH_PATTERNS,
): string | null {
  for (const re of patterns) {
    if (re.test(path)) return re.source;
  }
  return null;
}

/**
 * Build an Express middleware that blocks sensitive-path probes.
 * `arcis()` wires this on by default; users can opt out via
 * `arcis({ scannerPaths: false })`.
 */
export function scannerPathProtection(
  options: ScannerPathsOptions = {},
): RequestHandler {
  const { statusCode = 403, message = 'Access denied.' } = options;
  const patterns = options.patterns ?? SENSITIVE_PATH_PATTERNS;

  return (req: Request, res: Response, next: NextFunction) => {
    const matched = detectSensitivePath(req.path, patterns);
    if (matched === null) {
      return next();
    }
    (req as unknown as Record<string, unknown>).scannerPathHit = matched;
    res.status(statusCode).json({
      error: message,
      code: 'SECURITY_THREAT',
      vector: 'scanner-path',
    });
  };
}
